pkg/utils/etcdutil/etcdutil.go

```diff
@@ -196,72 +210,233 @@ func EtcdKVPutWithTTL(ctx context.Context, c *clientv3.Client, key string, value
-		return nil, nil, errs.ErrNewEtcdClient.Wrap(err).GenWithStackByCause()
-	}
-	httpClient := createHTTPClient(tlsConfig)
-	return client, httpClient, nil
-}
+const (
+	// etcdServerOfflineTimeout is the timeout for an unhealthy etcd endpoint to be offline from healthy checker.
+	etcdServerOfflineTimeout = 30 * time.Minute
+	// etcdServerDisconnectedTimeout is the timeout for an unhealthy etcd endpoint to be disconnected from healthy checker.
+	etcdServerDisconnectedTimeout = 1 * time.Minute
+)
 
-// createEtcdClientWithMultiEndpoint creates etcd v3 client.
-// Note: it will be used by micro service server and support multi etcd endpoints.
-// FIXME: But it cannot switch etcd endpoints as soon as possible when one of endpoints is with io hang.
-func createEtcdClientWithMultiEndpoint(tlsConfig *tls.Config, acUrls []url.URL) (*clientv3.Client, error) {
-	if len(acUrls) == 0 {
-		return nil, errs.ErrNewEtcdClient.FastGenByArgs("no available etcd address")
-	}
-	endpoints := make([]string, 0, len(acUrls))
-	for _, u := range acUrls {
-		endpoints = append(endpoints, u.String())
-	}
+func newClient(tlsConfig *tls.Config, endpoints ...string) (*clientv3.Client, error) {
 	lgc := zap.NewProductionConfig()
 	lgc.Encoding = log.ZapEncodingName
-	autoSyncInterval := defaultAutoSyncInterval
-	dialKeepAliveTime := defaultDialKeepAliveTime
-	dialKeepAliveTimeout := defaultDialKeepAliveTimeout
-	failpoint.Inject("autoSyncInterval", func() {
-		autoSyncInterval = 10 * time.Millisecond
-	})
-	failpoint.Inject("closeKeepAliveCheck", func() {
-		autoSyncInterval = 0
-		dialKeepAliveTime = 0
-		dialKeepAliveTimeout = 0
-	})
 	client, err := clientv3.New(clientv3.Config{
 		Endpoints:            endpoints,
 		DialTimeout:          defaultEtcdClientTimeout,
-		AutoSyncInterval:     autoSyncInterval,
 		TLS:                  tlsConfig,
 		LogConfig:            &lgc,
-		DialKeepAliveTime:    dialKeepAliveTime,
-		DialKeepAliveTimeout: dialKeepAliveTimeout,
+		DialKeepAliveTime:    defaultDialKeepAliveTime,
+		DialKeepAliveTimeout: defaultDialKeepAliveTimeout,
 	})
-	if err == nil {
-		log.Info("create etcd v3 client", zap.Strings("endpoints", endpoints))
+	return client, err
+}
+
+// CreateEtcdClient creates etcd v3 client with detecting endpoints.
+func CreateEtcdClient(tlsConfig *tls.Config, acURLs []url.URL) (*clientv3.Client, error) {
+	urls := make([]string, 0, len(acURLs))
+	for _, u := range acURLs {
+		urls = append(urls, u.String())
 	}
+	client, err := newClient(tlsConfig, urls...)
+	if err != nil {
+		return nil, err
+	}
+
+	tickerInterval := defaultDialKeepAliveTime
+	failpoint.Inject("fastTick", func() {
+		tickerInterval = 100 * time.Millisecond
+	})
+	failpoint.Inject("closeTick", func() {
+		tickerInterval = 0
+	})
+	if tickerInterval == 0 {
+		return client, err
+	}
+
+	checker := &healthyChecker{
+		tlsConfig: tlsConfig,
+	}
+	eps := syncUrls(client)
+	checker.update(eps)
+
+	// Create a goroutine to check the health of etcd endpoints periodically.
```

Comment:
Do we need to add sync.WatchGroup for these two goroutines?

Reply:
wait group? I passed in ctx and wg in a previous version and later found that client.ctx could control the exit of these two gotoutines.

Reply:
Yes, sync.WaitGroup. ok